syntax = "proto2";

message TwoIntsProto {
  required int64 x = 1;
  required int64 y = 2;
}

// Increment x.
message TwoIntsIncrementXRequest {
}

message TwoIntsIncrementXReply {
}

// Decrement y.
message TwoIntsDecrementYRequest {
}

message TwoIntsDecrementYReply {
}

// Get.
message TwoIntsGetRequest {
}

message TwoIntsGetReply {
  required int64 x = 1;
  required int64 y = 2;
}

// Request and reply.
message TwoIntsTxnRequest {
  oneof request {
    TwoIntsIncrementXRequest increment_x = 1;
    TwoIntsDecrementYRequest decrement_y = 2;
    TwoIntsGetRequest get = 3;
  }
}

message TwoIntsTxnReply {
  enum Result {
    COMMITTED = 0;
    ABORTED = 1;
  }

  required Result result = 1;
  oneof reply {
    TwoIntsIncrementXReply increment_x = 2;
    TwoIntsDecrementYReply decrement_y = 3;
    TwoIntsGetReply get = 4;
  }
}
